ledger-core/v2/runner/events.go

```diff
@@ -0,0 +1,230 @@
+
+type eventDeactivate struct {
+	parentRequestReference insolar.Reference
+	parentObjectReference  insolar.Reference
+}
+
+func (e eventDeactivate) ParentObjectReference() insolar.Reference {
+	return e.parentObjectReference
+}
+
+func (e eventDeactivate) ParentRequestReference() insolar.Reference {
+	return e.parentRequestReference
+}
+
+type eventRouteCall struct {
+	parentRequestReference insolar.Reference
+	parentObjectReference  insolar.Reference
+
+	method    string
+	arguments []byte
+	object    insolar.Reference
+	prototype insolar.Reference
+	immutable bool
+	saga      bool
+}
+
+func (e eventRouteCall) Saga() bool {
+	return e.saga
+}
+
+func (e eventRouteCall) Immutable() bool {
+	return e.immutable
+}
+
+func (e eventRouteCall) Prototype() insolar.Reference {
+	return e.prototype
+}
+
+func (e eventRouteCall) Object() insolar.Reference {
+	return e.object
+}
+
+func (e eventRouteCall) Arguments() []byte {
+	return e.arguments
+}
+
+func (e eventRouteCall) Method() string {
+	return e.method
+}
+
+func (e eventRouteCall) ParentObjectReference() insolar.Reference {
+	return e.parentObjectReference
+}
+
+func (e eventRouteCall) ParentRequestReference() insolar.Reference {
+	return e.parentRequestReference
+}
+
+type RPCRouteCallEvent interface {
+	RPCEvent
+
+	SetSaga(isSaga bool) RPCRouteCallEvent
+	SetImmutable(isImmutable bool) RPCRouteCallEvent
+}
+
+type RPCOutgoingConstructor interface {
+	ConstructOutgoing(transcript runner.Execution) record.OutgoingRequest
+}
+
+type RPCEventParentFunc func(parentRequest insolar.Reference, parentObject insolar.Reference) RPCEventBuilder
+
+type RPCEventBuilder interface {
+	Deactivate() RPCEvent
+	SaveAsChild(prototype insolar.Reference, constructor string, arguments []byte) RPCEvent
+	RouteCall(object insolar.Reference, prototype insolar.Reference, method string, arguments []byte) RPCRouteCallEvent
+	GetCode(code insolar.Reference) RPCEvent
+}
+
+func (e eventSaveAsChild) ConstructOutgoing(execution runner.Execution) record.OutgoingRequest {
+	return record.OutgoingRequest{
+		Caller: execution.Request.Callee,
+		Nonce:  0,
+
+		CallType:  record.CTSaveAsChild,
+		Base:      &e.parentObjectReference,
+		Prototype: &e.prototype,
+		Method:    e.constructor,
+		Arguments: e.arguments,
+
+		APIRequestID: "",
+		Reason:       e.parentRequestReference,
+	}
+}
+
+func (e eventRouteCall) ConstructOutgoing(execution runner.Execution) record.OutgoingRequest {
+	returnMode := record.ReturnResult
+	if e.saga {
+		returnMode = record.ReturnSaga
+	}
+
+	return record.OutgoingRequest{
+		Caller: execution.Request.Callee,
+		Nonce:  0,
+
+		CallType:   record.CTMethod,
+		Base:       &e.parentObjectReference,
+		Object:     &e.object,
+		Prototype:  &e.prototype,
+		Method:     e.method,
+		Arguments:  e.arguments,
+		Immutable:  e.immutable,
+		ReturnMode: returnMode,
+
+		APIRequestID: "",
+		Reason:       e.parentRequestReference,
+	}
+}
+
+func (e eventRouteCall) SetSaga(isSaga bool) RPCRouteCallEvent {
```

Comment:
this is method on non pointer receiver so all the modifications will not affect object. Is it intended?